src/us/kbase/workspace/database/provenance/ProvenanceAction.java

```diff
@@ -0,0 +1,744 @@
+	 * @return the outgoing arguments.
+	 */
+	public List<String> getOutgoingArgs() {
+		return getList(outgoingArgs);
+	}
+
+	/** Get information about any external data used in this provenance action.
+	 * @return the external data.
+	 */
+	public List<ExternalData> getExternalData() {
+		return getList(externalData);
+	}
+
+	/** Get information about any sub actions taken as part of this provenance action.
+	 * @return the sub actions.
+	 */
+	public List<SubAction> getSubActions() {
+		return getList(subActions);
+	}
+
+	/** Get any custom provenance information for this action.
+	 * @return the custom provenance information.
+	 */
+	public Map<String, String> getCustom() {
+		return custom == null ? Collections.emptyMap() : custom;
+	}
+	
+	/** Get the description of this provenance action.
+	 * @return the description, if present.
+	 */
+	public Optional<String> getDescription() {
+		return Optional.ofNullable(description);
+	}
+
+	/** Get the resolved workspace object refs from {@link #getWorkspaceObjects()}. Each item in
+	 * the standard list has a resolved entry in the resolved list, if the resolved list isn't
+	 * empty.
+	 * @return the resolved workspace objects or an empty list if there are no resolved objects
+	 * present.
+	 */
+	public List<String> getResolvedObjects() {
+		return getList(resolvedObjects);
+	}
+	
+	private <T> List<T> getList(final List<T> list) {
+		return list == null ? Collections.emptyList() : list;
+	}
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((caller == null) ? 0 : caller.hashCode());
+		result = prime * result + ((commandLine == null) ? 0 : commandLine.hashCode());
+		result = prime * result + ((custom == null) ? 0 : custom.hashCode());
+		result = prime * result + ((description == null) ? 0 : description.hashCode());
+		result = prime * result + ((externalData == null) ? 0 : externalData.hashCode());
+		result = prime * result + ((incomingArgs == null) ? 0 : incomingArgs.hashCode());
+		result = prime * result + ((method == null) ? 0 : method.hashCode());
+		result = prime * result + ((methodParameters == null) ? 0 : methodParameters.hashCode());
+		result = prime * result + ((outgoingArgs == null) ? 0 : outgoingArgs.hashCode());
+		result = prime * result + ((resolvedObjects == null) ? 0 : resolvedObjects.hashCode());
+		result = prime * result + ((script == null) ? 0 : script.hashCode());
+		result = prime * result + ((scriptVersion == null) ? 0 : scriptVersion.hashCode());
+		result = prime * result + ((service == null) ? 0 : service.hashCode());
+		result = prime * result + ((serviceVersion == null) ? 0 : serviceVersion.hashCode());
+		result = prime * result + ((subActions == null) ? 0 : subActions.hashCode());
+		result = prime * result + ((time == null) ? 0 : time.hashCode());
+		result = prime * result + ((wsobjs == null) ? 0 : wsobjs.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```

Comment:
This makes me itch for some sort of meta-object protocol that you could use to iterate through all the properties / attributes of the two objects, comparing them and exiting if a difference is found.

Reply:
The equals method is generated by Eclipse fwiw, so it's very low maintenance burden. I keep meaning to learn Lombok and have it generate hashCode and equals in bytecode but I never get around to it.